metricbeat/module/kafka/partition/partition.go

```diff
@@ -109,25 +111,31 @@ func (m *MetricSet) Fetch(r mb.ReporterV2) error {
 		}
 
 		for _, partition := range topic.Partitions {
-			// partition offsets can be queried from leader only
-			if broker.ID() != partition.Leader {
-				debugf("broker is not leader (broker=%v, leader=%v)", broker.ID(), partition.Leader)
-				continue
-			}
-
 			// collect offsets for all replicas
-			for _, id := range partition.Replicas {
-
-				// Get oldest and newest available offsets
-				offOldest, offNewest, offOK, err := queryOffsetRange(broker, id, topic.Name, partition.ID)
-
-				if !offOK {
-					if err == nil {
-						err = errFailQueryOffset
-					}
+			for _, replicaID := range partition.Replicas {
+				oldestPartitionOffsets := topicPartitionPartitionOldestOffsets[topic.Name][partition.ID]
+				if oldestPartitionOffsets == nil {
+					msg := fmt.Errorf("no oldest partition offsets defined (%v:%v)", topic.Name, partition.ID)
+					m.Logger().Warn(msg)
+					r.Error(msg)
+					continue
+				} else if oldestPartitionOffsets.Err != nil {
+					msg := fmt.Errorf("failed to query kafka partition (%v:%v) oldest offsets: %v",
+						topic.Name, partition.ID, oldestPartitionOffsets.Err)
+					m.Logger().Warn(msg)
+					r.Error(msg)
+					continue
+				}
 
-					msg := fmt.Errorf("Failed to query kafka partition (%v:%v) offsets: %v",
-						topic.Name, partition.ID, err)
+				newestPartitionOffsets := topicPartitionPartitionNewestOffsets[topic.Name][partition.ID]
+				if newestPartitionOffsets == nil {
+					msg := fmt.Errorf("no newest partition offsets defined (%v:%v)", topic.Name, partition.ID)
+					m.Logger().Warn(msg)
+					r.Error(msg)
+					continue
+				} else if newestPartitionOffsets.Err != nil {
+					msg := fmt.Errorf("failed to query kafka partition (%v:%v) newest offsets: %v",
+						topic.Name, partition.ID, newestPartitionOffsets.Err)
```

Comment:
Nit. Extract this common logic for oldest and newest offset to a method?

Reply:
Extracted.